extensions/socket/internal.m

```diff
@@ -0,0 +1,657 @@
+///
+/// Parameters:
+///  * None
+///
+/// Returns:
+///  * The number of connections to the socket
+///
+static int socket_connections(lua_State *L) {
+    LuaSkin *skin = [LuaSkin shared];
+    [skin checkArgs:LS_TUSERDATA, USERDATA_TAG, LS_TBREAK];
+
+    HSAsyncSocket* asyncSocket = getUserData(L, 1);
+    NSInteger connections;
+    if (asyncSocket.userData == SERVER) {
+        connections = asyncSocket.connectedSockets.count;
+    } else {
+        connections = asyncSocket.isConnected ? 1 : 0;
+    }
+
+    lua_pushinteger(L, connections);
+    return 1;
+}
+
+/// hs.socket:info() -> table
+/// Method
+/// Returns information on the [`hs.socket`](#new) instance
+///
+/// Parameters:
+///  * None
+///
+/// Returns:
+///  * A table containing the following keys:
+///   * connectedAddress - `string` (`sockaddr` struct)
+///   * connectedHost - `string`
+///   * connectedPort - `number`
+///   * isConnected - `boolean`
+///   * isDisconnected - `boolean`
+///   * isIPv4 - `boolean`
+///   * isIPv4Enabled - `boolean`
+///   * isIPv4PreferredOverIPv6 - `boolean`
+///   * isIPv6 - `boolean`
+///   * isIPv6Enabled - `boolean`
+///   * isSecure - `boolean`
+///   * localAddress - `string` (`sockaddr` struct)
+///   * localHost - `string`
+///   * localPort - `number`
+///   * timeout - `number`
+///   * userData - `string`
+///
+static int socket_info(lua_State *L) {
+    LuaSkin *skin = [LuaSkin shared];
+    [skin checkArgs:LS_TUSERDATA, USERDATA_TAG, LS_TBREAK];
+    HSAsyncSocket* asyncSocket = getUserData(L, 1);
+
+    NSDictionary *info = @{
+        @"connectedAddress" : asyncSocket.connectedAddress ?: @"",
+        @"connectedHost" : asyncSocket.connectedHost ?: @"",
+        @"connectedPort" : @(asyncSocket.connectedPort),
+        @"isConnected": @(asyncSocket.isConnected),
+        @"isDisconnected": @(asyncSocket.isDisconnected),
+        @"isIPv4": @(asyncSocket.isIPv4),
+        @"isIPv4Enabled": @(asyncSocket.isIPv4Enabled),
+        @"isIPv4PreferredOverIPv6": @(asyncSocket.isIPv4PreferredOverIPv6),
+        @"isIPv6": @(asyncSocket.isIPv6),
+        @"isIPv6Enabled": @(asyncSocket.isIPv6Enabled),
+        @"isSecure": @(asyncSocket.isSecure),
+        @"localAddress" : asyncSocket.localAddress ?: @"",
+        @"localHost" : asyncSocket.localHost ?: @"",
+        @"localPort" : @(asyncSocket.localPort),
+        @"timeout" : @(asyncSocket.timeout),
+        @"userData" : asyncSocket.userData ?: @"",
+    };
+
+    [skin pushNSObject:info];
+    return 1;
+}
+
+
+static int userdata_gc(lua_State *L) {
+    asyncSocketUserData *userData = lua_touserdata(L, 1);
+    HSAsyncSocket* asyncSocket = (__bridge_transfer HSAsyncSocket *)userData->asyncSocket;
+    userData->asyncSocket = nil;
+
+    [asyncSocket disconnect];
+    [asyncSocket setDelegate:nil delegateQueue:NULL];
+    asyncSocket.readCallback = [[LuaSkin shared] luaUnref:refTable ref:asyncSocket.readCallback];
```

Comment:
I think it would be a good idea fo also `luaUnref` the `writeCallback` and `connectCallback` since it's possible that `__gc` might be called while a socket is live and waiting for data/connect/whatever.

They're all initialised to `LUA_NOREF`, and it's a safe no-op to call `luaUnref` on a variable that's `LUA_NOREF` :)